default/python/AI/AIstate.py

```diff
@@ -154,7 +154,7 @@ def __init__(self, aggression):
         self.newlySplitFleets = {}
         self.militaryRating = 0
         self.shipCount = 4
-        self.misc = {}
+        self.misc = {}  # Keys: "observed_enemies"
```

Comment:
At the very least, there is the 'enemies_sighted' key as well.

Reply:
I prefer to store them as separate variables:  `observed_enemies=set()` . It is simple and clear when you read the code.

Reply:
> At the very least, there is the 'enemies_sighted' key as well.

I've now added the other keys, and also renamed this new one to observed_empires, to clarify the distinction from 'enemies_sighted'

> I prefer to store them as separate variables: observed_enemies=set() . It is simple and clear when you read the code.

Isn't that just a tradeoff of complexity in different places, at least to avoid making savegames incompatible, because if you make it an attribute like that then you need to update the version and __setstate__ method?  I think the current treatment is quite clear, and actually simpler.

Reply:
> at least to avoid making savegames incompatible

How many lines of code you need to add to __setstate__ method to remove redundant complexity?

- `setdefault` does not make code more readable.  You do some lazy initialization inside business logic, instead of putting it to logicless initialization code. 
- IDE type hinting does not work well with dict of type `str: object`

Reply:
I have to agree with Cj here. In general, I would prefer storing variables directly. If there are things which need to be grouped, a class or a namedtuple helps both with type hinting and error checks. 

Using dicts with hardcoded string keys is kind of dangerous especially when missing those keys is somewhat expected: It is extremely hard to track down a bug based on a wrong ```get``` call because we just use the default value and no error/warning will ever be thrown. 

For example ```observed_enemies``` vs ```enemies_sighted``` - who would notice the error if I used ```sighted_enemies``` instead which is grammatically correct and in line with the other naming scheme.

If we want to use dicts, setter/getter functions should be used to avoid these problems (and provide type hinting).

Also, the ```__setstate__``` method has to be changed once and will never have to be looked at again (and even if we notice a few versions later that there was some hard to fix savegame compatibility bug, we could simply raise the minimum required version). The attribute itself is worked with as long as it exists.